### ISSUE TITLE
(Help!) How would I put the perseus front-end on a meteor server?

### ISSUE THREAD

**Issue by jamesBaker361** [Khan/perseus#89]
I have been trying to run the perseus front-end and all of it's cool features and components and  widgets on a meteor server, but I'm yet to be in any way successful. I'm an amateur who does not completely understand how perseus works. What would i need to do to run a meteor server that also has Perseus working? Thanks!

**Comment by kevinbarabash** [comment]
@jamesBaker361 perseus uses React so I would probably start by making sure you can render React components in your meteor application.  Once you have that working you should be able to `require` the `perseus.js` from the `build` folder after running `make build`.

**Comment by jamesBaker361** [comment]
@kevinbarabash I followed your advice (thanks so much btw!) having run make build in the perseus folder and then run meteor run in my project folder, i get some weird errors:
`
While processing files with ecmascript (for target web.browser):
   perseus/lib/raphael.js:3820:55: Deleting local variable in strict mode
   (3820:55)
   
   While processing files with templating (for target web.browser):
   perseus/math-input/index.html:1: Can't set DOCTYPE here.  (Meteor sets
   <!DOCTYPE html> for you)
   perseus/react-components/template.html:1: Expected one of: <body>, <head>,
   <template>
   perseus/difftest.html:1: Can't set DOCTYPE here.  (Meteor sets <!DOCTYPE
   html> for you)
   perseus/index.html:1: Can't set DOCTYPE here.  (Meteor sets <!DOCTYPE html>
   for you)
   
   While minifying app stylesheet:
   app/perseus/stylesheets/local-only/khan-site.css:2962:12: property missing
   ':'
   
   While processing files with ecmascript (for target os.osx.x86_64):
   perseus/lib/raphael.js:3788:55: Deleting local variable in strict mode
   (3788:55)`

Do you have any idea as to what this could mean?

**Comment by kevinbarabash** [comment]
> While processing files with ecmascript (for target web.browser):
perseus/lib/raphael.js:3820:55: Deleting local variable in strict mode
(3820:55)

My guess would be that Meteor might add `"use strict"` directives to the Javascript files that it processes.  The culprit is:

```
R.ninja = function () {
        oldRaphael.was ? (win.Raphael = oldRaphael.is) : delete Raphael;
        return R;
    };
```

I would try changing `delete Raphael` to `delete win.Raphael`.

I'm not sure why Meteor is trying process these HTML files in the `perseus` tree.  If you're using React you should be requiring the `build/perseus.js`.  And then render content using a `Renderer` object.  See https://github.com/khan/perseus#external-dependencies for more details.

**Comment by jamesBaker361** [comment]
Yep! That worked! Thanks! Do you have any idea about what the "While minifying app stylesheet:
   app/perseus/stylesheets/local-only/khan-site.css:2962:12: property missing
   ':'" error could be caused by?

**Comment by jamesBaker361** [comment]
Nevermind I just changed minifier and that stopped being a problem.

**Comment by AAZZIOUI** [comment]
I am using meteor with react. When importing the perseus.js file I got errors like : 
`Uncaught TypeError: Cannot read property 'concat' of undefined    at perseus.js 42222`
more details:
` at http://localhost:3000/app/app.js?hash=add522ea7aff592e0223f34b9e856a67fd1af849:51105:40
    at Object.<anonymous> (http://localhost:3000/app/app.js?hash=add522ea7aff592e0223f34b9e856a67fd1af849:51593:6)
    at __webpack_require__ (http://localhost:3000/app/app.js?hash=add522ea7aff592e0223f34b9e856a67fd1af849:126:31)
    at Object.assign.i (http://localhost:3000/app/app.js?hash=add522ea7aff592e0223f34b9e856a67fd1af849:39608:4)
    at __webpack_require__ (http://localhost:3000/app/app.js?hash=add522ea7aff592e0223f34b9e856a67fd1af849:126:31)
    at Object.<anonymous> (http://localhost:3000/app/app.js?hash=add522ea7aff592e0223f34b9e856a67fd1af849:29531:4)
    at __webpack_require__ (http://localhost:3000/app/app.js?hash=add522ea7aff592e0223f34b9e856a67fd1af849:126:31)
    at Object.<anonymous> (http://localhost:3000/app/app.js?hash=add522ea7aff592e0223f34b9e856a67fd1af849:37970:21)
    at __webpack_require__ (http://localhost:3000/app/app.js?hash=add522ea7aff592e0223f34b9e856a67fd1af849:126:31)
    at Object.<anonymous> (http://localhost:3000/app/app.js?hash=add522ea7aff592e0223f34b9e856a67fd1af849:28290:25)(anonymous function) @ perseus.js:42222(anonymous function) @ perseus.js:42215__webpack_require__ @ perseus.js:34Object.assign.i @ perseus.js:31819__webpack_require__ @ perseus.js:34(anonymous function) @ perseus.js:23464__webpack_require__ @ perseus.js:34(anonymous function) @ perseus.js:30509__webpack_require__ @ perseus.js:34(anonymous function) @ perseus.js:22452__webpack_require__ @ perseus.js:34Object.assign.i @ perseus.js:28070__webpack_require__ @ perseus.js:34(anonymous function) @ perseus.js:21072__webpack_require__ @ perseus.js:34Object.assign.i @ perseus.js:5560__webpack_require__ @ perseus.js:34module.exports @ perseus.js:1446__webpack_require__ @ perseus.js:34module.exports @ perseus.js:21828__webpack_require__ @ perseus.js:34(anonymous function) @ perseus.js:4222__webpack_require__ @ perseus.js:34skipMathJax @ perseus.js:364__webpack_require__ @ perseus.js:34(anonymous function) @ perseus.js:87__webpack_require__ @ perseus.js:34(anonymous function) @ perseus.js:67__webpack_require__ @ perseus.js:34(anonymous function) @ perseus.js:54(anonymous function) @ perseus.js:15webpackUniversalModuleDefinition @ perseus.js:7meteorInstall.imports.build.perseus.js @ perseus.js:5fileEvaluate @ modules-runtime.js?hash=ab7e8d3…:191require @ modules-runtime.js?hash=ab7e8d3…:116Mp.import @ modules.js?hash=f72d430…:340meteorInstall.client.AppRender.jsx @ AppRender.jsx:1fileEvaluate @ modules-runtime.js?hash=ab7e8d3…:191require @ modules-runtime.js?hash=ab7e8d3…:116(anonymous function) @ app.js?hash=add522e…:59920`

First I made line 42222 in perseus.js as a comment to see what will happen, but another issue about i18n showed up and so on.

I am new to both meteor and react. I think I have installed all required dependencies and don't understand what's happening exactly. I hope to find a solution to import persues.js file without errors. Thanks

**Comment by ariabuckles** [comment]
`Cannot read property 'concat' of undefined` looks like you're missing underscore. Perseus depends on a few global libraries being loaded, which are listed in the [readme](https://github.com/Khan/perseus#external-dependencies)

**Comment by AAZZIOUI** [comment]
Thank you @ariabuckles  for your help, but In fact I think I have all the necessary dependencies as described in the readme, here are dependencies from my package.json :
`"dependencies": {
    "aphrodite": "^1.0.0",
    "bcrypt": "^0.8.7",
    "bootstrap": "^3.3.7",
    "classnames": "^2.2.5",
    "jquery": "^3.1.1",
    "jquery-validation": "^1.15.1",
    "meteor-node-stubs": "~0.2.0",
    "react": "^15.3.2",
    "react-addons-create-fragment": "^15.3.2",
    "react-addons-css-transition-group": "^15.3.2",
    "react-addons-pure-render-mixin": "^15.3.2",
    "react-bootstrap": "^0.30.6",
    "react-dom": "^15.3.2",
    "react-komposer": "^1.13.1",
    "react-router": "^3.0.0",
    "react-router-bootstrap": "^0.23.1",
    "underscore": "^1.8.3"
  }` 

I have even tried to add the following in the top of perseus.js :
`/****My personal changes:*****/
import 'react';
import 'underscore';
import 'aphrodite';
import 'classnames';
import 'jquery';
import 'react-addons-css-transition-group';
import 'react-addons-create-fragment';
import 'react-addons-pure-render-mixin';
/****************************/`

but in vain. I still have the same issue which starts at the level of this portion of code : 
`	mouseHookProps = $.event.mouseHooks ? $.event.mouseHooks.props : [],
		mouseEventProps = $.event.props`**_.concat_**`( mouseHookProps ),`

I have no idea on what is the problem I am facing here. Any suggestions would be appreciated. Thanks

**Comment by ariabuckles** [comment]
You specifically need those dependencies as global variables on window, not dependencies in package.json (though that might be how you get them onto window). So you probably want some file you import before Perseus that sets things like `window._ = require('underscore')` for each of those dependencies.

Sorry, I know this is a mess.

**Comment by AAZZIOUI** [comment]
Thanks @ariabuckles, I followed your suggestions & added these lines before importing perseus.js: 

`import { Meteor } from 'meteor/meteor';`
`import React from 'react';`
`import { render } from 'react-dom';`
`window._ = require("underscore");`
`window.React = require("react");`
`window.React.__internalReactDOM = require("react-dom");`
`window.React.__internalAddons.createFragment = require("react-addons-create-fragment");`
`window.React.__internalAddons.PureRenderMixin = require("react-addons-pure-render-mixin");`
`window.React.__internalAddons.CSSTransitionGroup = require("react-addons-css-transition-group");`
`window.jQuery = require('jquery');`
`require("classnames");`
`require("aphrodite");`
and then :
`import  '../imports/build/perseus.css';`
`import  '../imports/build/perseus.js';`

But still have the same issue. I've also tried many variations in the hope of getting it right but with no impact on the current problem. am I doing something wrong here? or missing s.th? 
Sorry
